src/main/java/org/wildfly/security/auth/provider/oauth2/OAuth2Util.java

```diff
@@ -0,0 +1,185 @@
+import org.wildfly.security._private.ElytronMessages;
+import org.wildfly.security.auth.server.RealmUnavailableException;
+import org.wildfly.security.authz.Attributes;
+import org.wildfly.security.authz.MapAttributes;
+import org.wildfly.security.util.ByteStringBuilder;
+import org.wildfly.security.util.CodePointIterator;
+
+import javax.json.Json;
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonString;
+import javax.json.JsonValue;
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLContext;
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UnsupportedEncodingException;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * An utility class providing static methods to connect to standard OAuth2 HTTP endpoints.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+class OAuth2Util {
+
+    /**
+     * Introspects an OAuth2 Access Token using a RFC-7662 compatible endpoint.
+     *
+     * @param tokenIntrospectionUrl an {@link URL} pointing to a RFC-7662 compatible endpoint
+     * @param clientId the identifier of a client within the OAUth2 Authorization Server
+     * @param clientSecret the secret of the client
+     * @param token the access token to introspect
+     * @param sslContext the ssl context
+     * @param hostnameVerifier the hostname verifier
+     * @return a @{JsonObject} representing the response from the introspection endpoint or null if
+     */
+    static JsonObject introspectAccessToken(URL tokenIntrospectionUrl, String clientId, String clientSecret, String token, SSLContext sslContext, HostnameVerifier hostnameVerifier) throws RealmUnavailableException {
+        Assert.checkNotNullParam("clientId", clientId);
+        Assert.checkNotNullParam("clientSecret", clientSecret);
+        Assert.checkNotNullParam("token", token);
+
+        HttpURLConnection connection = null;
+
+        try {
+            connection = openConnection(tokenIntrospectionUrl, sslContext, hostnameVerifier);
+
+            HashMap<String, String> parameters = new HashMap<>();
+
+            parameters.put("token", token);
+            parameters.put("token_type_hint", "access_token");
+
+            byte[] params = buildParameters(parameters);
+
+            connection.setDoOutput(true);
+            connection.setRequestMethod("POST");
+            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+            connection.setRequestProperty("Content-Length", String.valueOf(params.length));
+            connection.setRequestProperty("Authorization", "Basic " + CodePointIterator.ofString(clientId + ":" + clientSecret).asUtf8().base64Encode().drainToString());
+
+            connection.getOutputStream().write(params);
```

Comment:
This should use try-with-resources, because you're never closing the request AFAICT.